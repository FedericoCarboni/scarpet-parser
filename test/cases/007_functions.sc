// Function calls and definitions
// comment
func(param1, param2, outer(outer_param)) -> (
    null
);

// optional comma ,
func(1, 2,);
