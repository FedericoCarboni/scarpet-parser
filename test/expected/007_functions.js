/** @type {import('../../lib/Parser.js').Node} */
export const root = {
    kind: 'BinaryExpression',
    operator: ';',
    lvalue: {
        kind: 'BinaryExpression',
        operator: ';',
        lvalue: {
            kind: 'BinaryExpression',
            operator: '->',
            lvalue: {
                kind: 'FunctionExpression',
                name: 'func',
                params: [
                    {
                        kind: 'Variable',
                        name: 'param1',
                        start: {
                            offset: 50,
                            line: 2,
                            character: 5,
                        },
                        end: {
                            offset: 56,
                            line: 2,
                            character: 11,
                        },
                    },
                    {
                        kind: 'Variable',
                        name: 'param2',
                        start: {
                            offset: 58,
                            line: 2,
                            character: 13,
                        },
                        end: {
                            offset: 64,
                            line: 2,
                            character: 19,
                        },
                    },
                    {
                        kind: 'FunctionExpression',
                        name: 'outer',
                        params: [
                            {
                                kind: 'Variable',
                                name: 'outer_param',
                                start: {
                                    offset: 72,
                                    line: 2,
                                    character: 27,
                                },
                                end: {
                                    offset: 83,
                                    line: 2,
                                    character: 38,
                                },
                            },
                        ],
                        start: {
                            offset: 66,
                            line: 2,
                            character: 21,
                        },
                        end: {
                            offset: 84,
                            line: 2,
                            character: 39,
                        },
                    },
                ],
                start: {
                    offset: 45,
                    line: 2,
                    character: 0,
                },
                end: {
                    offset: 85,
                    line: 2,
                    character: 40,
                },
            },
            rvalue: {
                kind: 'ParenthesisedExpression',
                value: {
                    kind: 'Constant',
                    name: 'null',
                    start: {
                        offset: 95,
                        line: 3,
                        character: 4,
                    },
                    end: {
                        offset: 100,
                        line: 4,
                        character: 0,
                    },
                },
                start: {
                    offset: 89,
                    line: 2,
                    character: 44,
                },
                end: {
                    offset: 101,
                    line: 4,
                    character: 1,
                },
            },
            comment: '// Function calls and definitions\n// comment\n',
            start: {
                offset: 45,
                line: 2,
                character: 0,
            },
            end: {
                offset: 101,
                line: 4,
                character: 1,
            },
        },
        rvalue: {
            kind: 'FunctionExpression',
            name: 'func',
            params: [
                {
                    kind: 'NumberLiteral',
                    literal: '1',
                    value: 1n,
                    start: {
                        offset: 129,
                        line: 7,
                        character: 5,
                    },
                    end: {
                        offset: 130,
                        line: 7,
                        character: 6,
                    },
                },
                {
                    kind: 'NumberLiteral',
                    literal: '2',
                    value: 2n,
                    start: {
                        offset: 132,
                        line: 7,
                        character: 8,
                    },
                    end: {
                        offset: 133,
                        line: 7,
                        character: 9,
                    },
                },
                null,
            ],
            start: {
                offset: 124,
                line: 7,
                character: 0,
            },
            end: {
                offset: 135,
                line: 7,
                character: 11,
            },
        },
        comment: null,
        start: {
            offset: 45,
            line: 2,
            character: 0,
        },
        end: {
            offset: 135,
            line: 7,
            character: 11,
        },
    },
    rvalue: null,
    comment: '// optional comma ,\n',
    start: {
        offset: 45,
        line: 2,
        character: 0,
    },
    end: {
        offset: 136,
        line: 7,
        character: 12,
    },
};
