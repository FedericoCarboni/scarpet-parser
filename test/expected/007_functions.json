{
  "errors": [],
  "warnings": [],
  "root": {
    "kind": "BinaryExpression",
    "operator": ";",
    "lvalue": {
      "kind": "BinaryExpression",
      "operator": ";",
      "lvalue": {
        "kind": "BinaryExpression",
        "operator": "->",
        "lvalue": {
          "kind": "FunctionExpression",
          "name": "func",
          "params": [
            {
              "kind": "Variable",
              "name": "param1",
              "start": {
                "pos": 50,
                "row": 2,
                "col": 5
              },
              "end": {
                "pos": 56,
                "row": 2,
                "col": 11
              }
            },
            {
              "kind": "Variable",
              "name": "param2",
              "start": {
                "pos": 58,
                "row": 2,
                "col": 13
              },
              "end": {
                "pos": 64,
                "row": 2,
                "col": 19
              }
            },
            {
              "kind": "FunctionExpression",
              "name": "outer",
              "params": [
                {
                  "kind": "Variable",
                  "name": "outer_param",
                  "start": {
                    "pos": 72,
                    "row": 2,
                    "col": 27
                  },
                  "end": {
                    "pos": 83,
                    "row": 2,
                    "col": 38
                  }
                }
              ],
              "start": {
                "pos": 66,
                "row": 2,
                "col": 21
              },
              "end": {
                "pos": 84,
                "row": 2,
                "col": 39
              }
            }
          ],
          "start": {
            "pos": 45,
            "row": 2,
            "col": 0
          },
          "end": {
            "pos": 85,
            "row": 2,
            "col": 40
          }
        },
        "rvalue": {
          "kind": "ParenthesisedExpression",
          "value": {
            "kind": "Constant",
            "name": "null",
            "start": {
              "pos": 95,
              "row": 3,
              "col": 4
            },
            "end": {
              "pos": 100,
              "row": 4,
              "col": 0
            }
          },
          "start": {
            "pos": 89,
            "row": 2,
            "col": 44
          },
          "end": {
            "pos": 101,
            "row": 4,
            "col": 1
          }
        },
        "comment": "// Function calls and definitions\n// comment\n",
        "start": {
          "pos": 45,
          "row": 2,
          "col": 0
        },
        "end": {
          "pos": 101,
          "row": 4,
          "col": 1
        }
      },
      "rvalue": {
        "kind": "FunctionExpression",
        "name": "func",
        "params": [
          {
            "kind": "NumberLiteral",
            "literal": "1",
            "value": "1",
            "start": {
              "pos": 129,
              "row": 7,
              "col": 5
            },
            "end": {
              "pos": 130,
              "row": 7,
              "col": 6
            }
          },
          {
            "kind": "NumberLiteral",
            "literal": "2",
            "value": "2",
            "start": {
              "pos": 132,
              "row": 7,
              "col": 8
            },
            "end": {
              "pos": 133,
              "row": 7,
              "col": 9
            }
          },
          null
        ],
        "start": {
          "pos": 124,
          "row": 7,
          "col": 0
        },
        "end": {
          "pos": 135,
          "row": 7,
          "col": 11
        }
      },
      "comment": null,
      "start": {
        "pos": 45,
        "row": 2,
        "col": 0
      },
      "end": {
        "pos": 135,
        "row": 7,
        "col": 11
      }
    },
    "rvalue": null,
    "comment": "// optional comma ,\n",
    "start": {
      "pos": 45,
      "row": 2,
      "col": 0
    },
    "end": {
      "pos": 136,
      "row": 7,
      "col": 12
    }
  }
}